apps/blockchain/lib/blockchain/transaction.ex

```diff
@@ -362,23 +365,36 @@ defmodule Blockchain.Transaction do
   def pay_and_refund_gas(state, sender, trx, total_refund, block_header) do
     # Eq.(74)
     # Eq.(75)
+
     state
     |> Account.add_wei(sender, total_refund * trx.gas_price)
     |> Account.add_wei(block_header.beneficiary, (trx.gas_limit - total_refund) * trx.gas_price)
   end
 
-  @spec clean_up_accounts_marked_for_destruction(EVM.state(), EVM.SubState.t(), Block.Header.t()) ::
+  @spec clean_up_accounts_marked_for_destruction(EVM.state(), EVM.SubState.t(), Header.t()) ::
           EVM.state()
-  defp clean_up_accounts_marked_for_destruction(state, sub_state, block_header) do
+  defp clean_up_accounts_marked_for_destruction(state, sub_state, header) do
     Enum.reduce(sub_state.selfdestruct_list, state, fn address, new_state ->
-      if Header.mined_by?(block_header, address) do
-        Account.reset_account(new_state, address)
+      if Header.mined_by?(header, address) do
+        Account.reset_account(state, address)
       else
         Account.del_account(new_state, address)
       end
     end)
   end
 
+  @spec maybe_reset_coinbase(EVM.state(), EVM.SubState.t(), Header.t()) :: EVM.state()
+  defp maybe_reset_coinbase(state, sub_state, header) do
```

Comment:
In the current implementation, we're resetting the coinbase account in `clean_up_accounts_marked_for_destruction` if the coinbase was self-destructed. But receipts expect state with the coinbase deleted not resetted. So we should reset coinbase after setting `root_hash` of the receipt.

Currently this method is not used, it will be used after turning on receipts validation in `HollisticValidity`